cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,805 @@
+
+  rmm::device_uvector<edge_type_t> unique_types(0, handle.get_stream());
+  rmm::device_uvector<edge_t> unique_type_counts(0, handle.get_stream());
+
+  if constexpr (multi_gpu) {
+    auto& comm                 = handle.get_comms();
+    auto const comm_size       = comm.get_size();
+    auto& major_comm           = handle.get_subcomm(cugraph::partition_manager::major_comm_name());
+    auto const major_comm_size = major_comm.get_size();
+    auto& minor_comm           = handle.get_subcomm(cugraph::partition_manager::minor_comm_name());
+    auto const minor_comm_size = minor_comm.get_size();
+
+    auto [gpu_ids, edge_types] =
+      cugraph::extract_transform_e(
+        handle,
+        graph_view,
+        cugraph::edge_src_dummy_property_t{}.view(),
+        cugraph::edge_dst_dummy_property_t{}.view(),
+        view_concat(edge_id_view, edge_type_view),
+        cuda::proclaim_return_type<thrust::optional<thrust::tuple<int, edge_type_t>>>(
+          [key_func =
+             cugraph::detail::compute_gpu_id_from_ext_edge_t<edge_t>{
+               comm_size,
+               major_comm_size,
+               minor_comm_size}] __device__(auto,
+                                            auto,
+                                            thrust::nullopt_t,
+                                            thrust::nullopt_t,
+                                            thrust::tuple<edge_t, edge_type_t> id_and_type) {
+            return thrust::optional<thrust::tuple<int, edge_type_t>>{thrust::make_tuple(
+              key_func(thrust::get<0>(id_and_type)), thrust::get<1>(id_and_type))};
+          }));
+
+    auto type_and_gpu_id_pair_begin =
+      thrust::make_zip_iterator(thrust::make_tuple(edge_types.begin(), gpu_ids.begin()));
+
+    thrust::sort(handle.get_thrust_policy(),
+                 type_and_gpu_id_pair_begin,
+                 type_and_gpu_id_pair_begin + edge_types.size());
+
+    auto nr_unique_paris = thrust::count_if(
+      handle.get_thrust_policy(),
+      thrust::make_counting_iterator(size_t{0}),
+      thrust::make_counting_iterator(edge_types.size()),
+      detail::is_first_in_run_t<decltype(type_and_gpu_id_pair_begin)>{type_and_gpu_id_pair_begin});
+
+    auto unique_pairs_buffer = cugraph::allocate_dataframe_buffer<
+      typename thrust::iterator_traits<decltype(type_and_gpu_id_pair_begin)>::value_type>(
+      nr_unique_paris, handle.get_stream());
+
+    rmm::device_uvector<edge_t> unique_pair_counts(nr_unique_paris, handle.get_stream());
+
+    thrust::reduce_by_key(handle.get_thrust_policy(),
+                          type_and_gpu_id_pair_begin,
+                          type_and_gpu_id_pair_begin + edge_types.size(),
+                          thrust::make_constant_iterator(size_t{1}),
+                          cugraph::get_dataframe_buffer_begin(unique_pairs_buffer),
+                          unique_pair_counts.begin());
```

Comment:
Same here. thrust::copy_if + binary search will be more efficient.

Reply:
Changed.